src/main/java/io/tiledb/spark/DataSourceOptions.java

```diff
@@ -0,0 +1,133 @@
+package io.tiledb.spark;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * An immutable string-to-string map in which keys are case-insensitive. This is used to represent
+ * data source options.
+ *
+ * <p>Each data source implementation can define its own options and teach its users how to set
+ * them. Spark doesn't have any restrictions about what options a data source should or should not
+ * have. Instead Spark defines some standard options that data sources can optionally adopt. It's
+ * possible that some options are very common and many data sources use them. However different data
+ * sources may define the common options(key and meaning) differently, which is quite confusing to
+ * end users.
+ */
+public class DataSourceOptions {
```

Comment:
Is this custom code?

Reply:
This class is taken from the Spark2 API and is not present in Spark3. I copy-pasted it in an effort to keep changes to a minimum.

Reply:
Please add a license attribution.

Reply:
Done!